scripts/gulpfiles/release_notes.js

```diff
@@ -0,0 +1,201 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @fileoverview Gulp scripts for creating release notes.
+ */
+
+var gulp = require('gulp');
+var fetch = require('node-fetch');
+
+const baseUrl = "https://api.github.com/graphql";
+
+async function generateReleaseNotes(done) {
+  const goalDate = Date.parse('2021-03-01');
+  const githubToken = process.argv[4];
+  const headers = {
+    "Content-Type": "application/json",
+    Authorization: "bearer " + githubToken,
+  };
+
+
+  var note = `{Insert greeting},
+
+The {Insert Month and Year} release is out! In this release:
+
+{List major features in both core and samples}
+
+🎉 **Kudos** 🎉
+{Thank forums contributors and issue filers}
+
+**Third Party Plugins**
+{List third party plugins that have been released in the last quarter}
+
+**Core contributions**
+{Reorganize into thank-yous and check for missed people}
+${await getContributors('blockly', 'develop', goalDate, headers)}
+
+**Samples contributions**
+{Reorganize into thank-yous and check for missed people}
+${await getContributors('blockly-samples', 'master', goalDate, headers)}
+
+🔨 **Breaking Changes in core** 🔨 
+${await getBreakingChanges(headers)}
+
+Full release notes are below. As always, we welcome bug reports and pull requests!
+
+Cheers,
+{Your Name}
+
+**Issues Closed**:
+${await getClosedIssues(headers)}
+
+**Pull Requests**:
+${await getMergedPrs(headers)}`;
+
+  console.log(note);
+
+  done();
+}
+
+async function getContributors(repo, baseRef, goalDate, headers) {
+  const collaborators = await getCollaborators(repo, headers);
+
+  let cursor = '';
+  const contributors = new Map();
+  do {
+    try {
+      const response = await fetch(baseUrl, {
+        method: "POST",
+        headers: headers,
+        body: makeContributorQuery(repo, baseRef, cursor),
+      });
+      const json = await response.json();
+      const pullRequestData = json.data.repository.pullRequests;
+      const pulls = pullRequestData.nodes;
+      cursor = pullRequestData.pageInfo.endCursor;
+    
+      for (let i = 0; i < pulls.length; i++) {
+        const pull = pulls[i];
+        const createdDate = Date.parse(pull.createdAt);
+        if (createdDate < goalDate) {
+          cursor = '';
+          break;
+        }
+
+        if (collaborators.has(pull.author.login)) {
+          continue;
+        }
+        let contributions = contributors.get(pull.author.login);
+        if (!contributions) {
+          contributions = [];
+          contributors.set(pull.author.login, contributions);
+        }
+        contributions.push({
+          title: pull.title,
+          url: pull.url,
+          number: pull.number
+        })
+      }
+
+    } catch (error) {
```

Comment:
Wrapping such a big block of code in `try`/`catch` seems like  code smell.  Do you really want to continue iterating if _anything_ in the block throws?

* If the idea is just to ensure that errors are correctly reported to/by gulp, then I'd move the `try`/`catch` to the top-level Gulp task (in this case `generateReleaseNotes`—but note that you will want to report to Gulp that the task failed, and the easiest way to do that is to have the promise reject, which can be achieved simply by _not handling the exception_ in the top-level `async function`.  (Do verify this works as you expect, obviously.)
* If the intention is to retry failed queries then I'd try to reduce the scope of the `try` block to contain only the statement(s) that you are expecting might fail and whose failure should not unwind the stack.  Add a `continue` to the `catch` block if you want to skip the post-query processing that would otherwise follow.

Reply:
Thanks for mentioning this, it should definitely be reorganized. My intention is to retry failed queries, but I'm confused about what you mean by:
> Add a `continue` to the `catch` block if you want to skip the post-query processing that would otherwise follow.

Could you give a code sample for what this would look like? I think that would really help clear things up for me :D

Reply:
Sure.  A very simple example might look like this:

```JS
while (true) {
  let input = prepareInput();
  let output;
  try {
    output = thisMightFail(input);
  } catch (err) {
    handleError();
    continue;
  }
  postprocess(output);
}
```

This will cause exceptions thrown by `prepareInput` or `postprocess` (which we do not expect to throw) to terminate the loop, while exceptions from `thisMightFail` (which we do expect to throw) are dealt with—say, by logging them.  The `continue` ensures that `postprocess` only gets called if `thisMightFail` succeeds.

Compare to this version which also only runs `postprocess` if `thisMightFail` succeeds, but will swallow unexpected exceptions from `postprocess` along with the expected ones from `thisMightFail`:

```JS
while (true) {
  let input = prepareInput();
  let output;
  try {
    output = thisMightFail(input);
    postprocess(output);
  } catch (err) {
    handleError();
  }
}
```

Reply:
> Also do you have a recommendation for how best to retry failed queries? Should everything go in a loop?

I'm not an expert in this area but I'd say your general approach here looks pretty good.  The things I would check are:

* What happens if the first request fails?  If I am reading correctly, it looks like `cursor` won't get set and the loop will terminate without retrying.

* What happens if a subsequent request fails?  It looks like cursor will be unchanged and so it will retry at the same position on the next loop iteration (provided nothing after `cursor = …` throws).

Note that you may want to put an upper limit on the number of failures.  This could be global (give up after N failed requests in total), or per-request (give up if any individual request fails N times); in the latter case putting the request in an inner `do {request} while (!success && !tooManyAttempts)` loop could be helpful.

Reply:
(There's also the small matter of what to do if the request succeeds but the returned data is invalid / truncated.  Retrying the same request may or may not help.)